protographic/tests/operations/message-builder.test.ts

```diff
@@ -0,0 +1,386 @@
+
+      const message = buildMessageFromSelectionSet('UserResponse', userSelection.selectionSet, userType, typeInfo);
+
+      expect(message.name).toBe('UserResponse');
+      expect(message.fields.id).toBeDefined();
+      expect(message.fields.profile).toBeDefined();
+
+      // Should have nested message for profile (now with simple name)
+      expect(message.nested).toBeDefined();
+      expect(message.nested!.Profile).toBeDefined();
+    });
+
+    test('should use field number manager', () => {
+      const schema = buildSchema(`
+        type Query {
+          user: User
+        }
+        
+        type User {
+          id: ID!
+          name: String
+        }
+      `);
+
+      const query = parse(`
+        query GetUser {
+          user {
+            id
+            name
+          }
+        }
+      `);
+
+      const operation = query.definitions[0];
+      if (operation.kind !== 'OperationDefinition' || !operation.selectionSet) {
+        throw new Error('Invalid operation');
+      }
+
+      const userSelection = operation.selectionSet.selections[0];
+      if (userSelection.kind !== 'Field' || !userSelection.selectionSet) {
+        throw new Error('Invalid selection');
+      }
+
+      const typeInfo = new TypeInfo(schema);
+      const userType = schema.getType('User') as GraphQLObjectType;
+      const manager = createFieldNumberManager();
+
+      const message = buildMessageFromSelectionSet('UserResponse', userSelection.selectionSet, userType, typeInfo, {
+        fieldNumberManager: manager,
+      });
+
+      expect(manager.getFieldNumber('UserResponse', 'id')).toBeDefined();
+      expect(manager.getFieldNumber('UserResponse', 'name')).toBeDefined();
+    });
+
+    test('should handle list fields', () => {
+      const schema = buildSchema(`
+        type Query {
+          users: [User!]!
+        }
+        
+        type User {
+          id: ID!
+          name: String
+        }
+      `);
+
+      const query = parse(`
+        query GetUsers {
+          users {
+            id
+            name
+          }
+        }
+      `);
+
+      const operation = query.definitions[0];
+      if (operation.kind !== 'OperationDefinition' || !operation.selectionSet) {
+        throw new Error('Invalid operation');
+      }
+
+      const usersSelection = operation.selectionSet.selections[0];
+      if (usersSelection.kind !== 'Field' || !usersSelection.selectionSet) {
+        throw new Error('Invalid selection');
+      }
+
+      const typeInfo = new TypeInfo(schema);
+      const userType = schema.getType('User') as GraphQLObjectType;
+
+      const message = buildMessageFromSelectionSet('UsersResponse', usersSelection.selectionSet, userType, typeInfo);
+
+      expect(message.name).toBe('UsersResponse');
+      expect(message.fields.id).toBeDefined();
+      expect(message.fields.name).toBeDefined();
+    });
+
+    test('should handle field aliases', () => {
```

Comment:
Do we also have tests for root aliases or is this forbidden? Do we ensure that only a single operation can be inside a document?

Reply:
Going to add validation to ensure single operation per document.
root aliases will mean we cannot reverse when converting the proto msg back to GraphQL.